compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -447,389 +677,212 @@ extChEnv :: MonadEnv m ph
   -> Type
   -- ^ The result type of the new choice.
   -> m ()
-extChEnv tc ch self this arg upd typ =
-  let substUpd sExp tExp aExp = substituteTm (createExprSubst [(self,sExp),(this,tExp),(arg,aExp)]) upd
-  in getEnv >>= \case
-    EnvVG{} -> error "Impossible: extChEnv is not used in the value gathering phase"
-    env@EnvCG{..} -> putEnv env{_envcgchs = HM.insert (UpdChoice tc ch) (ChoiceData self this arg substUpd typ) _envcgchs}
-    env@EnvS{..} -> putEnv env{_envschs = HM.insert (UpdChoice tc ch) (ChoiceData self this arg substUpd typ) _envschs}
+extChEnv tc ch self this arg upd typ = do
+  env <- getEnv
+  putEnv $ setEnvChoices (HM.insert (UpdChoice tc ch) (ChoiceData self this arg upd typ) $ envChoices env) env
 
 -- | Extend the environment with a list of new data type definitions.
-extDatsEnv :: MonadEnv m ph
+extDatsEnv :: (IsPhase ph, MonadEnv m ph)
   => HM.HashMap TypeConName DefDataType
   -- ^ A hashmap of the data constructor names, with their corresponding definitions.
   -> m ()
-extDatsEnv hmap = getEnv >>= \case
-    env@EnvVG{..} -> putEnv env{_envvgdats = hmap `HM.union` _envvgdats}
-    env@EnvCG{..} -> putEnv env{_envcgdats = hmap `HM.union` _envcgdats}
-    env@EnvS{..} -> putEnv env{_envsdats = hmap `HM.union` _envsdats}
-
--- | Extend the environment with a new contract id, and the variable to which
--- the fetched contract is bound.
-extCidEnv :: MonadEnv m ph
-  => Expr
+extDatsEnv hmap = do
+  env <- getEnv
+  putEnv $ setEnvDats (hmap `HM.union` envDats env) env
+
+-- | Extend the environment with a refreshed contract id, and the variable to
+-- which the fetched contract is bound. Returns a substitution, mapping the
+-- given contract id, to the refreshed one.
+-- While it might seem counter intuitive, the function only refreshes the
+-- contract id on its first encounter. The reason is that it needs to be able to
+-- keep track of old bindings.
+-- Note that instead of overwriting old bindings, the function creates a new
+-- synonym between the old and new binding.
+extCidEnv :: (IsPhase ph, MonadEnv m ph)
+  => Bool
+  -- ^ Flag denoting whether the contract id should be refreshed.
+  -- Note that even with the flag on, contract id are only refreshed on their
+  -- first encounter.
+  -> Expr
   -- ^ The contract id expression.
   -> ExprVarName
   -- ^ The variable name to which the fetched contract is bound.
+  -> m ExprSubst
+extCidEnv b exp var = do
+  case exp of
+    -- Filter out any bindings to `_`.
+    EVar (ExprVarName "ds2") -> return emptyExprSubst
```

Comment:
That seems sketchy. Is that what GHC uses? Why do we need to filter this out, i.e., what harm is there in not filtering it out?

Reply:
You're right, this was an issue before, when multiple updates were bound to an underscore. But as it has been fixed in the meantime, this can be removed :)